### ISSUE TITLE
`ML backend: fix CRR bug, add cross-validation, and make the model swappable (LR → RF → XGBoost)`

### ISSUE THREAD

**Issue by dhanasai2** [Arnav-Singh-5080/IPL-Win-Probability-Prediction#92]
**Labels:** `enhancement` · `ML / backend` · `GSSoC 2026` · `good first issue`

---

Hey @Arnav-Singh-5080 , I spent the past couple of days going through the codebase (specifically `application.py`) and genuinely love what you've built here. The glassmorphism UI is clean, and the ball-by-ball feature engineering pipeline is solid for a first version.

That said, I dug into the ML backend and found a few things worth fixing, some are silent bugs, some are architectural gaps that'll make future contributions harder. I'd love to tackle this as my GSSoC contribution if you're open to it. Here's everything I found:

---

### What's currently going on (the problems)

**1. There's a data corruption bug in the CRR calculation**

Around line ~900, the code does this to avoid a division-by-zero:

```python
df['over'] = df['over'].replace(0, 0.1)
```

The problem is `over=0` represents the entire first over — all 6 balls of it. Replacing it with `0.1` means the CRR for those balls gets inflated by 10×. Across the training set that's roughly 450+ corrupted rows (6 balls × ~75 matches), and those rows go straight into the model without any warning. The fix is just a couple of lines of proper ball-count math, but it's silently skewing predictions right now.

---

**2. The model trains on 100% of the data no evaluation at all**

The training function does `pipe.fit(X, y)` on the full dataset and that's it. No train/test split, no cross-validation, no logged accuracy or AUC score anywhere. So honestly we have no idea how well the model actually performs, it could be overfitting, it could be great, we just don't know. For a project positioning itself as a production analytics platform, this feels like a gap worth closing.

---

**3. `pipe.pkl` exists in the repo but is never actually loaded**

There's a serialized model file sitting in the root directory, but `application.py` completely ignores it and retrains from scratch every single time the app starts. The `@st.cache_resource` decorator helps within a session, but on cold starts (which happen constantly on Streamlit Cloud) the app re-runs the full Pandas merge + feature engineering + `pipe.fit()` on ~170K rows. It's unnecessary compute and makes the app noticeably slow to load.

---

**4. The model is hardcoded and can't be swapped**

I noticed the README explicitly lists "Random Forest / XGBoost toggle" as an open good-first-issue, but the current architecture doesn't support it — the `LogisticRegression` estimator is hardcoded deep inside the pipeline. Swapping it right now means editing the core training function, which isn't great for contributors. A small refactor would open this up completely.

---

### What I want to fix

I'm proposing one focused PR that covers all four of these. Here's the plan:

1. **Fix the CRR bug** — replace the `over=0` workaround with proper computation: `balls_bowled = (over * 6) + ball`, guard with `balls_bowled > 0`, compute CRR cleanly. No more corrupted rows.

2. **Add proper evaluation** — split the data 80/20, run 5-fold cross-validation, and log accuracy + AUC-ROC + Brier score so we actually know what we're working with.

3. **Make the estimator swappable** — refactor `train_model()` to accept a `model_name` arg (`"logistic"`, `"random_forest"`, `"xgboost"`). This directly resolves the open issue in the README. Each estimator gets sensible pre-tuned defaults so results are good out of the box.

4. **Actually use `pipe.pkl`** — save the trained pipeline to disk after first run, load it on subsequent boots. Cold start goes from ~15s to under 1s. Cache-busts automatically when the model choice changes.

Rough shape of the new API, just to show I've thought this through:

```python
@st.cache_resource
def get_model(model_name: str = "logistic") -> Pipeline:
    pkl_path = f"models/{model_name}.pkl"
    if os.path.exists(pkl_path):
        return joblib.load(pkl_path)
    pipe = build_and_train(model_name)
    joblib.dump(pipe, pkl_path)
    return pipe
```

I'll keep the PR scoped to the backend — no UI changes beyond wiring in the model selector if you want it. Everything will be tested locally before I open the PR.

---

### My checklist

- [x] Fix CRR calculation (zero-division bug, corrupted rows)
- [x] Add 80/20 split + 5-fold CV with logged metrics
- [x] Refactor pipeline to support logistic / random_forest / xgboost via flag
- [x] Wire up model persistence so cold starts are fast
- [x] Add docstrings to all modified functions
- [x] Test locally, make sure the Streamlit app still runs cleanly before PR

---

Could you assign this to me? I want to make sure nobody else is working on the same thing before I start. I'll have a draft PR up within a week and will keep this thread updated.

Happy to scope it differently if you'd prefer, for example, splitting the bug fix and the model refactor into two separate issues. Just let me know what works better for the project.

Thanks for building this, genuinely one of the more interesting GSSoC projects I've come across this season.

— @dhanasai2
GSSoC 2026 contributor

**Comment by dhanasai2** [comment]
Hi @Arnav-Singh-5080, the PR linked to this issue appears completed and merged, but the contribution hasn't reflected on the GSSoC leaderboard yet after 48+ hours. Could you please verify if everything is properly tagged from the repository side? Thanks!
